# ParSE-seq

Interpreting the clinical significance of putative splice-altering variants outside 2-base pair canonical splice sites remains difficult without functional studies. We developed Parallel Splice Effect Sequencing (ParSE-seq), a multiplexed minigene-based assay, to test variant effects on RNA splicing quantified by high-throughput sequencing. This method allows us to explore the landscape of splice-altering variants in the arrhythmia-linked gene SCN5A, the main sodium ion channel of the heart. We study hundreds of variants that allow us to calibrate our assay to ClinGen specifications, interpret VUS and CI variants, correlate with SpliceAI and other in silico predictors, and reveal occasional shortfalls of cDNA-based patch-clamp assays. 

Here we present we present our 2 part code. First, we process barcode frequencies, assign barcodes to unique constructs, calculate Percent Splice In associated with each barcode, and then merge on variant. Next,  we explore this data and perform a variety of studies that constitute the main figures of the manuscript. We include additional dataframes and outputs that are generated in Unix and Python in additional folders. All DNA and RNA sequencing data is available at the NCBI Sequence Read Archive (NIH BioProject accession #1106089). All clinical classifications arising from this work were uploaded to ClinVar. A full manuscript describing and contextualizing this analysis is named 'ParSE-seq: a calibrated multiplexed assay to facilitate the clinical classification of putative splice-altering variants' with PubMed 39333091 (O'Neill et al, Nature Communications 2024). Please write to moneill22@bwh.harvard.edu or andrew.m.glazer@vumc.org with questions!  
